docs: transforms updates (#69037)
metabase/metabase: docs/data-studio/transforms/python-transforms.md

---
title: Python transforms
summary: Use Python to wrangle your data in Metabase and write the results back to your database.
---

# Python transforms

> Python transforms require the **Transforms add-on**

You can write a Python script to transform the data from your database, and write the result back into the database as another table.

For general information about Metabase transforms, see [Transforms](transforms.md).

## How Python transforms work

Python-based transforms require a dedicated [Python execution environment](#set-up-a-python-runner).

- In your Metabase, you write a Python script that returns a `pandas` DataFrame and uses one or more tables from your database.

> - When Metabase runs the transform, a new Python execution environment is spun up. Python transforms run in a separate, isolated environment—not on your Metabase instance.

Reviewer: ```suggestion
- Metabase will run your transform in a new Python execution environment (not on your Metabase instance).
```

- Metabase securely copies your source data to your Python environment and makes it available as pandas DataFrames.
- The Python environment executes your Python script _in memory_.

> - The Python environment saves the resulting DataFrame as a file.

Reviewer: ```suggestion
- The Python environment executes your Python script _in memory_ and saves the resulting DataFrame as a file.
```

- Your Metabase instance reads the file and writes the results to a new table in your database.
- The new table is synced to Metabase.
- On subsequent transform runs, your database will overwrite that table with the updated results (updates are not incremental).

Only one Python transform can run in the execution environment at once.

## Set up a Python runner

To execute Python transforms, you'll need a _Python runner_ - a dedicated environment for running Python code.

> If you're on Metabase Cloud, all you need it to add the **Transforms add-on**, and you'll be good to go.

Reviewer: ```suggestion
If you're on Metabase Cloud, get the **Transforms add-on**, and you're good to go.
```

If you're self-hosting Metabase, you'll need to set up a self-hosted Python runner, see [Python runner](python-runner.md).

## Create a Python transform

> To write Python transforms, you'll need to [Set up the Python runner](#set-up-a-python-runner). Once you've set up the runner:

Reviewer: ```suggestion
Once you've [set up the Python runner](#set-up-a-python-runner):
```

1. Go to **Data studio > Transforms**.

2. Click **+New** and select **Python transform**

> 3. Select a database that has the data you want to transform. See [Databases that support transforms](transforms.md#databases-that-support-transforms).

Reviewer: ```suggestion
3. Select a database that has the data you want to transform. See [databases that support transforms](transforms.md#databases-that-support-transforms).
```

7. Once you're done with you code, click **Save** in the top right corner.

8. Select a target schema for your transform and enter a name for the target table. Metabase will write the results of the transform into this table.

   You can only write back to the same database as you chose for the transform source.

9. Optionally, make your transform incremental. See [Incremental Python transforms](#incremental-python-transforms)
10. Optionally, assign tags to your transforms. Tags are used by [jobs](jobs-and-runs.md) to run transforms on schedule.

## Tips for writing Python transforms

- Metabase will automatically add `import common` to the code of your Python transform. This imports the [Python library](#common-python-library). You can use Python library for reusable functions and classes.
- A Python transform must have a function `transform()` that returns a single `pandas` DataFrame.
- You can use aliases to include tables from your database as DataFrames inside the `transform()` function. The tables will _only_ be available in the transform function. Other functions won't have access to the tables.

> - Only `pandas` will be imported by default, but you can import [certain other packages](#available-python-packages). You can also use functions from the [common library](#common-python-library).

Reviewer: ```suggestion
- Only `pandas` will be imported by default, but you can import [other packages](#available-python-packages). You can also use functions from the [common library](#common-python-library).
```

Due to security considerations for the Python execution environment, you won't be able to install or import any other packages.

You can also write your own functions and add them to the [common Python library](#common-python-library) in Metabase, which will be available across all your Python transforms.

## Common Python library

If you have functions or classes you'd like to reuse across multiple transforms, you can add them to the common library that will be available across all Python transforms.

To add things to the common Python library:

1. Go to **Data studio > Transforms**.
2. Scroll to the very bottom of the transforms list and click on **Python library**.

3. Add a Python function or class.

   Functions in this library can't access any data in your database.

To use functions or classes from your Python library:

1. Metabase will automatically add `import common` to your transform's code. The `common` here refers to the Python library. You can click on `common` in the editor to navigate to the library.

2. You can reference functions or classes from the common library in your code like `common.manifest_kittens()`.

## Incremental Python transforms

> By default, on every transform run after the first one, Metabase will process all the data in all input tables, then drop the existing target table, and create a new table with the processed data. You can tell Metabase to only write **new** data to your target table by marking you transform as incremental.

Reviewer: ```suggestion
By default, Metabase will process all the data in all input tables, drop the existing target table (if one exists), and create a new table with the processed data. You can tell Metabase to only write **new** data to your target table by marking you transform as incremental.
```

### Prerequisites for incremental transforms

You data has to have certain structure for incremental transforms to work. See [Prerequisites for incremental transforms](./transforms-overview.md#prerequisites-for-incremental-transforms).

### Make a Python transform incremental

To make a Python transform incremental:

1. Go to the transform's page in **Data studio > Transforms**.
2. Switch to **Settings** tab.
3. In **Column to check for new values**, select the column that Metabase should check to determine which values are new. See [Prerequisites for incremental transforms](./transforms-overview.md#prerequisites-for-incremental-transforms) for more information on the requirements for that column

>    You have to select the column from the list of the columns of the _input_ tables. Note: this is different from [Query transforms](./query-transforms.md), where you select an _output_ column as column to check for new values.

Reviewer: ```suggestion
   Unlike [Query transforms](./query-transforms.md), where you select an _output_ column as the column to check for new values, with Python transforms, you have to select a column from the _input_ tables as the column to check for new values.
```
